EIPS/eip-4200.md

```diff
@@ -0,0 +1,121 @@
+
+In great many cases control flow is actually static and there is no need for any dynamic behaviour, though not every use case can be solved by them.
+
+There are various ways to reduce the need for dynamic jumps, some examples:
+1. With native support for functions / subroutines
+2. A "return to caller" instruction
+3. A "switch-case" table with dynamic indexing
+
+We are not attempting to solve these, but decided to only introduce a minimal feature set to allow compilers to decide which is the most adequate option for a given use case. We expect they would use `RJUMP`/`RJUMPI` almost exclusively, with the exception of returning to the caller would keep using `JUMP`.
+
+This functionality does not preclude the EVM from introducing other forms of control flow later on. In fact we think `RJUMP`/`RJUMPI` can efficiently co-exists with a higher-level declaration of functions, where static relative jumps should be used for intra-function control flow.
+
+The main resulting benefit of these instruction is reduced gas cost (both at deploy and execution time) and better analysis properties.
+
+## Specification
+
+We introduce two new instructions on the same block number [EIP-3540](./eip-3540.md) is activated on:
+1. `RJUMP` (0x5c)
+2. `RJUMPI` (0x5d)
+
+If the code is legacy bytecode, both of these instructions result in an *exceptional halt*. (*Note: This means no change to behaviour.*)
+
+> TODO: Should we ignore what happens in the legacy case, or it is nicer being explicit as above?
+
+If the code is valid EOF1:
+1. `RJUMP relative_offset` (0x5c), sets the `PC` to `PC_post_instruction + relative_offset`.
+2. `RJUMPI relative_offset` (0x5d), pops a value (`condition`) from the stack, and sets the `PC` to `PC_post_instruction + ((condition == 0) ? 0 : relative_offset)`.
+
+The immediate argument `relative_offset` is encoded as a 16-bit **signed** (twos-complement) big-endian value. Under `PC_post_instruction` we mean the `PC` position after the entire immediate value.
+
+We also extend the validation algorithm of [EIP-3670](./eip-3670.md) to verify that each `RJUMP`/`RJUMPI` has a `relative_offset` pointing to an instruction. This means it cannot point to an immediate data of `PUSHn`/`RJUMP`/`RJUMPI`. It is allowed to point to a `JUMPDEST`, but is not required to.
+
+Because the destinations are validated upfront, the cost of these instructions are less than their dynamic counterparts: `RJUMP` should cost 6, and `RJUMPI` should cost 8. This is a reduction of 2 gas, compared to `JUMP` and `JUMPI`.
+
+## Rationale
+
+### Relative addressing
+
+We chose relative addressing in order to support code which is moveable. This also means it can be injected. A technique used by Optimism's validator was to inject code like `PUSHn PC ADD JUMPI` to get around this problem.
```

Comment:
```suggestion
We chose relative addressing in order to support code which is moveable. This also means it can be injected. A technique seen used prior to this EIP to achieve this same goal was to inject code like `PUSHn PC ADD JUMPI`.
```
Recommend remaining neutral and not mentioning specific dapps when possible (avoids people thinking you are advertising).